### PR TITLE
Add library reference handling

### PR DESCRIPTION
**Pull request by rwalton-arm**
### Description

Add a mechanism to hydrate library dependencies.

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: mbed_project/mbed_program.py [hunk 1]
```diff
@@ -54,14 +61,23 @@ def from_remote_url(cls, url: str, dst_path: Path) -> "MbedProgram":
                 "to an empty directory."
             )
 
-        logger.info(f"Cloning Mbed program from URL {url}")
+        print(f"Cloning Mbed program from URL '{url}'.")
```
Comment (acabarbaye) [inline mbed_project/mbed_program.py:64]: should it actually be `print` and not a call to the logger?

File: mbed_project/mbed_program.py [hunk 2]
```diff
@@ -104,9 +120,49 @@ def from_existing_local_program_directory(cls, dir_path: Path) -> "MbedProgram":
...
+        """Clone all external dependencies defined in .lib files.
+
+        This function will recursively clone all unresolved dependencies until all are resolved.
+
+        Raises:
+            VersionControlError: The git clone failed.
+        """
+        for lib in self._find_unresolved_lib_references():
+            git_ref = lib.get_git_reference()
+            print(f"Resolving library reference {git_ref.repo_url}.")
```
Comment (acabarbaye) [inline mbed_project/mbed_program.py:136]: shouldn't it be a call to the logger?

File: mbed_project/mbed_program.py (hunk 1 above)
Comment (rwalton-arm) [reply mbed_project/mbed_program.py:64]: Not sure. I want these messages to appear regardless of log level. Which is why I went with `print` here.

File: mbed_project/mbed_program.py (hunk 2 above)
Comment (rwalton-arm) [reply mbed_project/mbed_program.py:136]: Not sure. I want these messages to appear regardless of log level so I went with `print`.

File: mbed_project/mbed_program.py [hunk 3]
```diff
@@ -104,9 +120,49 @@ def from_existing_local_program_directory(cls, dir_path: Path) -> "MbedProgram":
...
+            try:
+                repo = git.Repo.clone_from(git_ref.repo_url, str(lib.source_code_path))
```
Comment (acabarbaye) [inline mbed_project/mbed_program.py:138]: Could this bit about resolving a single library be in a separate method?

Comment (rwalton-arm) [reply mbed_project/mbed_program.py:138]: Yeah, could do. I'll see how it looks.

File: mbed_project/mbed_program.py (hunk 1 above)
Comment (acabarbaye) [reply mbed_project/mbed_program.py:64]: the problem is that if a person redirects logs to a file, these messages would not appear unless I am mistaken.

Comment (rwalton-arm) [reply mbed_project/mbed_program.py:64]: I was trying to make a distinction between user facing messages and developer facing log messages if that makes sense. That's why I left the other logger calls in there. Maybe it's not worth it and I should send everything to the logger? `print` has a `file` arg which defaults to `stdout`, but anyone would have to set that specifically if they were redirecting to a file. I'm wondering if we should set the default log level to "info" or something if we go the logger route? I think it's important these messages are shown to the user in the "default" case without having to pass verbosity flags.

Comment (rwalton-arm) [reply mbed_project/mbed_program.py]: fe70a74

File: mbed_project/mbed_program.py (hunk 2 above)
Comment (rwalton-arm) [reply mbed_project/mbed_program.py]: fe70a74

File: mbed_project/mbed_program.py (hunk 3 above)
Comment (rwalton-arm) [reply mbed_project/mbed_program.py]: fe70a74

File: mbed_project/_internal/project_data.py [hunk 4]
```diff
@@ -87,16 +117,19 @@ def from_existing(cls, root_path: Path) -> "MbedProgramData":
...
         if not config.exists():
             logger.info("This program does not contain an mbed_app.json config file.")
             config = None
 
-        mbed_file = root_path / ".mbed"
-        if not mbed_file.exists():
-            raise ValueError(f"An Mbed program does not exist at {root_path}")
+        mbed_os_file = root_path / MBED_OS_REFERENCE_FILE_NAME
+        if not mbed_os_file.exists():
+            raise ValueError("This path does not contain an mbed-os.lib, which is required for mbed programs.")
```
Comment (rosiemiller2) [inline mbed_project/_internal/project_data.py:127]: Is it worth mentioning here what the mbed-os.lib file is for? I imagine that if someone has hit this error then they probably don't know.

Comment (rwalton-arm) [reply mbed_project/_internal/project_data.py:127]: Yes, makes sense. I'll add some more words.

Comment (rosiemiller2) [reply mbed_project/_internal/project_data.py:127]: I take it back - its explained later on.

Comment (rosiemiller2) [reply mbed_project/_internal/project_data.py:127]: It's up to you. It looks like the error will be incorporated into the longer explanation you provide in `MbedProgram.from_remote_url()`, but I guess being as explicit as possible can't hurt.

Comment (rwalton-arm) [reply mbed_project/_internal/project_data.py:127]: Oh yeah, so it is. It's been a long, long weekend and I had forgotten what my own error messages were.

**Comment by acabarbaye** [comment]
👍

**Pull request merged**